### PR TITLE
Fixed wing controller.

### PR DESCRIPTION
**Pull request by arjo129**
This PR adds a fixed wing controller for the zephyr. It implements a simple attitude controller that uses the `mavros_msgs` `AttitudeTarget.msg` for being commanded. The zephyr is a bit of an oddity when it comes to fix wings as it doesn't have control surfaces for controlling yaw. Instead one can get the aircraft to roll in one direction and thus achieve yaw control. I believe we can add this feature in a bit later as its a relatively trivial feature to add.

## To Test
:warning: This PR pulls in a new dependency on `mavros_msgs` you are likely going to have to do the whole `rosdep` drill again.

Start an empty ignition gazebo world in one window:
```
ign gazebo
```
Spawn a fixed wing zephyr in a separate terminal:
```
ros2 launch mbzirc_ign spawn.launch.py name:=zephyr world:=empty_platform model:=mbzirc_fixed_wing type:=uav x:=0 y:=0 z:=4
```
Publish a `mavros_msgs/msg/AttitudeTarget` on `/zephyr/cmd/attitude` with a `thrust` of 1000.
image

Start the simulation and you should see the plane try to level out:
plane_leveling_out

## TODO
* [x] Write unit tests
* [x] Document plugin

### CODE REVIEW

**Comment by iche033** [comment]
I'm testing this in the simple_demo world that now has 4ms step size and ran into a crash when spawning the fixed wing UAV. I think https://github.com/ignitionrobotics/ign-gazebo/pull/1366 should fix it. Can you take a look @arjo129?

**Comment by iche033** [comment]
cool I can get it to work after the lift drag fix by following these steps:

1. start the simulation paused
2. spawn the fixed wing UAV at 4m above ground
3. send the thrust command
4. play the simulation

Let's worry about the launching / take off strategy after we have more info

I made some follow-up changes in https://github.com/osrf/mbzirc/pull/54 to support mounting sensors

**Comment by arjo129** [comment]
> I'm testing this in the simple_demo world that now has 4ms step size and ran into a crash when spawning the fixed wing UAV. I think [ignitionrobotics/ign-gazebo#1366](https://github.com/ignitionrobotics/ign-gazebo/pull/1366) should fix it. Can you take a look @arjo129?
> 
> To reproduce:
> 
> ```
> # note: sim needs to be running and needs the -r flag to reproduce
> ros2 launch ros_ign_gazebo ign_gazebo.launch.py ign_args:="-v 4 -r simple_demo.sdf"
> 
> ros2 launch mbzirc_ign spawn.launch.py name:=zephyr world:=simple_demo model:=mbzirc_fixed_wing type:=uav x:=0 y:=0 z:=4
> ```

I was not able to reproduce the crash but the fix seems to be legitemate.

**Comment by iche033** [comment]
> I was not able to reproduce the crash but the fix seems to be legitemate.

hmm weird. Don't know if it's timing issue or not but it happens to me every time I run it.

**Comment by arjo129** [comment]
We seem to have a flaky tests.

**Comment by iche033** [comment]
ign-gazebo 6.6.0 debs are out, which contains the lift drag patch. The last build passed. I just triggered another one to make sure.

**Comment by iche033** [comment]
hmm all tests passed but I think there is a crash on exit with this msg `corrupted double-linked list`

**Comment by arjo129** [comment]
Yeah I've been experiencing that locally also after the fix. It seems to be flaky (and there are different stacktraces between different runs). I'm not too sure whats causing it but likely its going to take some time to hunt down the root cause. Do we want to first disable this test then revisit it again in a future iteration?

File: mbzirc_ign/CMakeLists.txt [hunk 1]
```diff
@@ -86,14 +90,14 @@ if(BUILD_TESTING)
...
+    test_spawn_uav_battery test_game_logic test_spawn_usv_maxspeed test_fixed_wing)
     ament_add_gtest(
       ${TEST_TARGET}
       test/${TEST_TARGET}.cc
     )
     target_include_directories(${TEST_TARGET}
-      PRIVATE ${CMAKE_BINARY_DIR} ${CMAKE_BINARY_DIR}/proto )
-    target_link_libraries(${TEST_TARGET} ignition-gazebo6::ignition-gazebo6)
+      PRIVATE ${CMAKE_BINARY_DIR} ${CMAKE_BINARY_DIR}/proto mavros_msgs)
+    target_link_libraries(${TEST_TARGET} ignition-gazebo6::ignition-gazebo6 rclcpp::rclcpp ${mavros_msgs_LIBRARIES})
```
Comment (mjcarroll) [inline mbzirc_ign/CMakeLists.txt:100]:
```suggestion
    target_link_libraries(${TEST_TARGET} ignition-gazebo${IGN_GAZEBO_VER}::ignition-gazebo${IGN_GAZEBO_VER} rclcpp::rclcpp ${mavros_msgs_LIBRARIES})
```

File: mbzirc_ign/src/FixedWingController.cc [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+};
+
+FixedWingControllerPlugin::FixedWingControllerPlugin() :
+  dataPtr(std::make_unique<FixedWingControllerPrivate>())
+{
+}
+
+FixedWingControllerPlugin::~FixedWingControllerPlugin()
+{
+}
```
Comment (mjcarroll) [inline mbzirc_ign/src/FixedWingController.cc:137]:
```suggestion
FixedWingControllerPlugin::~FixedWingControllerPlugin() = default;
```

**Review by mjcarroll — COMMENTED** [review]
A few cleanup nits

**Comment by iche033** [comment]
here are the 2 crashes with backtraces that I caught. Both related to the use of `rclcpp::Node`:

<details><summary>backtrace 1 (spin)</summary>

```
0x00007ffff7b453a1 in rclcpp::Executor::add_node(std::shared_ptr<rclcpp::node_interfaces::NodeBaseInterface>, bool) ()
   from /opt/ros/galactic/lib/librclcpp.so
(gdb) bt
#0  0x00007ffff7b453a1 in rclcpp::Executor::add_node(std::shared_ptr<rclcpp::node_interfaces::NodeBaseInterface>, bool) ()
   from /opt/ros/galactic/lib/librclcpp.so
#1  0x00007ffff7b41752 in rclcpp::Executor::spin_node_some(std::shared_ptr<rclcpp::node_interfaces::NodeBaseInterface>) ()
   from /opt/ros/galactic/lib/librclcpp.so
#2  0x00007ffff7b49920 in rclcpp::spin_some(std::shared_ptr<rclcpp::node_interfaces::NodeBaseInterface>) ()
   from /opt/ros/galactic/lib/librclcpp.so
#3  0x00007ffff7b49b00 in rclcpp::spin_some(std::shared_ptr<rclcpp::Node>) () from /opt/ros/galactic/lib/librclcpp.so
#4  0x00005555555f3c27 in MBZIRCTestFixture_FixedWingController_Test::TestBody()::{lambda(ignition::gazebo::v6::UpdateInfo const&, ignition::gazebo::v6::EntityComponentManager&)#2}::operator()(ignition::gazebo::v6::UpdateInfo const&, ignition::gazebo::v6::EntityComponentManager&) const ()
#5  0x00005555555f584f in std::_Function_handler<void (ignition::gazebo::v6::UpdateInfo const&, ignition::gazebo::v6::EntityComponentManager&), MBZIRCTestFixture_FixedWingController_Test::TestBody()::{lambda(ignition::gazebo::v6::UpdateInfo const&, ignition::gazebo::v6::EntityComponentManager&)#2}>::_M_invoke(std::_Any_data const&, ignition::gazebo::v6::UpdateInfo const&, ignition::gazebo::v6::EntityComponentManager&) ()
#6  0x000055555562b1db in std::function<void (ignition::gazebo::v6::UpdateInfo const&, ignition::gazebo::v6::EntityComponentManager&)>::operator()(ignition::gazebo::v6::UpdateInfo const&, ignition::gazebo::v6::EntityComponentManager&) const ()
#7  0x000055555560a13f in MBZIRCTestFixture::OnPreUpdate(std::function<void (ignition::gazebo::v6::UpdateInfo const&, ignition::gazebo::v6::EntityComponentManager&)>)::{lambda(ignition::gazebo::v6::UpdateInfo const&, ignition::gazebo::v6::EntityComponentManager&)#1}::operator()(ignition::gazebo::v6::UpdateInfo const&, ignition::gazebo::v6::EntityComponentManager&) const ()
#8  0x00005555556305f8 in std::_Function_handler<void (ignition::gazebo::v6::UpdateInfo const&, ignition::gazebo::v6::EntityComponentManager&), MBZIRCTestFixture::OnPreUpdate(std::function<void (ignition::gazebo::v6::UpdateInfo const&, ignition::gazebo::v6::EntityComponentManager&)>)::{lambda(ignition::gazebo::v6::UpdateInfo const&, ignition::gazebo::v6::EntityComponentManager&)#1}>::_M_invoke(std::_Any_data const&, ignition::gazebo::v6::UpdateInfo const&, ignition::gazebo::v6::EntityComponentManager&) ()
#9  0x00007ffff7ee7a42 in ignition::gazebo::v6::SimulationRunner::UpdateSystems (this=this@entry=0x555555b56380)
    at /home/ian/code/ign_f_ws/src/ign-gazebo/src/SimulationRunner.cc:581
#10 0x00007ffff7eeff6b in ignition::gazebo::v6::SimulationRunner::Step (this=this@entry=0x555555b56380, _info=...)
    at /home/ian/code/ign_f_ws/src/ign-gazebo/src/SimulationRunner.cc:854
#11 0x00007ffff7ef07eb in ignition::gazebo::v6::SimulationRunner::Run (this=0x555555b56380, _iterations=<optimized out>)
    at /home/ian/code/ign_f_ws/src/ign-gazebo/src/SimulationRunner.cc:809
#12 0x00007ffff7ee0e94 in ignition::gazebo::v6::ServerPrivate::Run (this=0x555555ad3330, _iterations=<optimized out>, _cond=...)
    at /usr/include/c++/9/bits/unique_ptr.h:360
#13 0x00007ffff6d3bde4 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#14 0x00007ffff792c609 in start_thread (arg=<optimized out>) at pthread_create.c:477
#15 0x00007ffff6a2b293 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95

```
</details>

<details><summary>backtrace 2 (destructor)</summary>

```
#0  futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x555555a5d760) at ../sysdeps/nptl/futex-internal.h:183
#1  __pthread_cond_wait_common (abstime=0x0, clockid=0, mutex=0x555555a5d710, cond=0x555555a5d738) at pthread_cond_wait.c:508
#2  __pthread_cond_wait (cond=0x555555a5d738, mutex=0x555555a5d710) at pthread_cond_wait.c:638
#3  0x00007ffff5082c2d in ddsrt_cond_wait () from /opt/ros/galactic/lib/x86_64-linux-gnu/libddsc.so.0
#4  0x00007ffff50363fb in rtps_stop () from /opt/ros/galactic/lib/x86_64-linux-gnu/libddsc.so.0
#5  0x00007ffff50606aa in ?? () from /opt/ros/galactic/lib/x86_64-linux-gnu/libddsc.so.0
#6  0x00007ffff5065789 in dds_delete_impl_pinned () from /opt/ros/galactic/lib/x86_64-linux-gnu/libddsc.so.0
#7  0x00007ffff50659ab in ?? () from /opt/ros/galactic/lib/x86_64-linux-gnu/libddsc.so.0
#8  0x00007ffff510cff0 in ?? () from /opt/ros/galactic/lib/librmw_cyclonedds_cpp.so
#9  0x00007ffff510d088 in ?? () from /opt/ros/galactic/lib/librmw_cyclonedds_cpp.so
#10 0x00007ffff510d399 in rmw_destroy_node () from /opt/ros/galactic/lib/librmw_cyclonedds_cpp.so
#11 0x00007ffff6e758e5 in rcl_node_fini () from /opt/ros/galactic/lib/librcl.so
#12 0x00007ffff7b724de in ?? () from /opt/ros/galactic/lib/librclcpp.so
#13 0x00007ffff7b72685 in ?? () from /opt/ros/galactic/lib/librclcpp.so
#14 0x00007ffff7b72cdc in rclcpp::node_interfaces::NodeBase::~NodeBase() () from /opt/ros/galactic/lib/librclcpp.so
#15 0x00007ffff7b72e2d in rclcpp::node_interfaces::NodeBase::~NodeBase() () from /opt/ros/galactic/lib/librclcpp.so
#16 0x00007ffff7b68210 in rclcpp::Node::~Node() () from /opt/ros/galactic/lib/librclcpp.so
#17 0x00007fffac4e791d in void __gnu_cxx::new_allocator<rclcpp::Node>::destroy<rclcpp::Node>(rclcpp::Node*) () from /home/ian/code/mbzirc_ws/install/lib/libFixedWingController.so
#18 0x00007fffac4e1d87 in void std::allocator_traits<std::allocator<rclcpp::Node> >::destroy<rclcpp::Node>(std::allocator<rclcpp::Node>&, rclcpp::Node*) () from /home/ian/code/mbzirc_ws/install/lib/libFixedWingController.so
#19 0x00007fffac4de2a1 in std::_Sp_counted_ptr_inplace<rclcpp::Node, std::allocator<rclcpp::Node>, (__gnu_cxx::_Lock_policy)2>::_M_dispose() () from /home/ian/code/mbzirc_ws/install/lib/libFixedWingController.so
#20 0x000055555562c1bc in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release() ()
#21 0x00007fffac45aff7 in std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count() () from /home/ian/code/mbzirc_ws/install/lib/libFixedWingController.so
#22 0x00007fffac45a0e8 in std::__shared_ptr<rclcpp::Node, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr() () from /home/ian/code/mbzirc_ws/install/lib/libFixedWingController.so
#23 0x00007fffac45a108 in std::shared_ptr<rclcpp::Node>::~shared_ptr() () from /home/ian/code/mbzirc_ws/install/lib/libFixedWingController.so
#24 0x00007fffac464396 in mbzirc::FixedWingControllerPrivate::~FixedWingControllerPrivate() () from /home/ian/code/mbzirc_ws/install/lib/libFixedWingController.so
#25 0x00007fffac46442c in std::default_delete<mbzirc::FixedWingControllerPrivate>::operator()(mbzirc::FixedWingControllerPrivate*) const () from /home/ian/code/mbzirc_ws/install/lib/libFixedWingController.so
#26 0x00007fffac45fb02 in std::unique_ptr<mbzirc::FixedWingControllerPrivate, std::default_delete<mbzirc::FixedWingControllerPrivate> >::~unique_ptr() () from /home/ian/code/mbzirc_ws/install/lib/libFixedWingController.so
#27 0x00007fffac452fb0 in mbzirc::FixedWingControllerPlugin::~FixedWingControllerPlugin() () from /home/ian/code/mbzirc_ws/install/lib/libFixedWingController.so
#28 0x00007fffac452fdc in mbzirc::FixedWingControllerPlugin::~FixedWingControllerPlugin() () from /home/ian/code/mbzirc_ws/install/lib/--Type <RET> for more, q to quit, c to continue without paging--
libFixedWingController.so
#29 0x00007fffac4659d0 in ignition::plugin::detail::Registrar<mbzirc::FixedWingControllerPlugin, ignition::gazebo::v6::System, ignition::gazebo::v6::ISystemConfigure, ignition::gazebo::v6::ISystemPreUpdate>::MakeInfo()::{lambda(void*)#2}::operator()(void*) const () from /home/ian/code/mbzirc_ws/install/lib/libFixedWingController.so
#30 0x00007fffac47774f in std::_Function_handler<void (void*), ignition::plugin::detail::Registrar<mbzirc::FixedWingControllerPlugin, ignition::gazebo::v6::System, ignition::gazebo::v6::ISystemConfigure, ignition::gazebo::v6::ISystemPreUpdate>::MakeInfo()::{lambda(void*)#2}>::_M_invoke(std::_Any_data const&, void*&&) () from /home/ian/code/mbzirc_ws/install/lib/libFixedWingController.so
#31 0x00007ffff683df96 in std::function<void (void*)>::operator()(void*) const (__args#0=<optimized out>, this=0x555558acfd38) at /usr/include/c++/9/bits/std_function.h:683
#32 ignition::plugin::PluginWithDlHandle::~PluginWithDlHandle (this=0x555558acfd20, __in_chrg=<optimized out>) at /home/ian/code/ign_f_ws/src/ign-plugin/core/src/Plugin.cc:64
```
</details>

File: mbzirc_ign/src/FixedWingController.cc (hunk 2 above)
Comment (iche033) [reply mbzirc_ign/src/FixedWingController.cc]: [0a87671](https://github.com/osrf/mbzirc/pull/51/commits/0a87671f56fa892ba1a9b56277971a6034bc8b41)

File: mbzirc_ign/CMakeLists.txt (hunk 1 above)
Comment (iche033) [reply mbzirc_ign/CMakeLists.txt]: [0a87671](https://github.com/osrf/mbzirc/pull/51/commits/0a87671f56fa892ba1a9b56277971a6034bc8b41)

File: mbzirc_ign/models/mbzirc_fixed_wing_base/model.sdf [hunk 3]
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" ?>
 <sdf version="1.6">
-  <model name="zephyr_delta_wing">
+  <model name="zephyr_ delta_wing">
```
[the remark this answers is not in the document]
Comment (iche033) [reply mbzirc_ign/models/mbzirc_fixed_wing_base/model.sdf]: probably not. I updated it. [0a87671](https://github.com/osrf/mbzirc/pull/51/commits/0a87671f56fa892ba1a9b56277971a6034bc8b41)

File: mbzirc_ign/src/FixedWingController.cc [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+  /// \brief Pitch control PID
+  public: math::PID pitchControl;
+
+  /// \brief Target Roll control PID
+  public: double targetRoll{0};
+
+  /// \brief Target Pitch control setpoint
+  public: double targetPitch{0};
+
+  /// \brief Yaw control PID
```
[the remark this answers is not in the document]
Comment (iche033) [reply mbzirc_ign/src/FixedWingController.cc]: [0a87671](https://github.com/osrf/mbzirc/pull/51/commits/0a87671f56fa892ba1a9b56277971a6034bc8b41)

**Comment by iche033** [comment]
I left the fixed UAV flying in the coast environment for 20mins and haven't run into a crash yet. @mjcarroll, Do you know if those `rclcpp::Node` crashes are harmful or not? If it only affects the test, I'll just disable the test for now.


I noticed a couple of other things:
* max bounds are incorrect - fixed in [17e0e82](https://github.com/osrf/mbzirc/pull/51/commits/17e0e829424919d5706882ba04c90d039450df06)
* UAV eventually curves to the right (right click on zephyr and and Follow to see this). I think it just needs a yaw controller - we can comeback to this and see if we need to provide this or not.

**Comment by iche033** [comment]
> here are the 2 crashes with backtraces that I caught. Both related to the use of rclcpp::Node:

should be fixed by #55 

I'll merge this when CI is green

**Review by iche033 — APPROVED** [review]

**Review by mjcarroll — APPROVED** [review]

**Pull request merged**